File: holoeasy-core/src/main/java/org/holoeasy/hologram/Hologram.java
package org.holoeasy.hologram;

import com.github.retrooper.packetevents.PacketEvents;
import com.github.retrooper.packetevents.wrapper.play.server.WrapperPlayServerEntityTeleport;
import io.github.retrooper.packetevents.util.SpigotConversionUtil;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.holoeasy.HoloEasy;
import org.holoeasy.line.*;
import org.holoeasy.pool.IHologramPool;
import org.holoeasy.pool.KeyAlreadyExistsException;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Function;

public class Hologram {
    private final HoloEasy lib;
    private final PrivateConfig pvt;
    private final UUID id = UUID.randomUUID();
    private Location location;
    private final List<Line<?>> lines = new CopyOnWriteArrayList<>();
    private boolean loaded = false;

    public Hologram(@NotNull HoloEasy lib, @NotNull Location location, ShowEvent showEvent, HideEvent hideEvent) {
        this.lib = lib;
        this.location = location;
        this.pvt = new PrivateConfig(this, showEvent, hideEvent);
    }

    public Hologram(@NotNull HoloEasy lib, @NotNull Location location) {
        this(lib, location, null, null);
    }

    public @NotNull HoloEasy getLib() {
        return lib;
    }

    @ApiStatus.Internal
    public @NotNull PrivateConfig getPvt() {
        return pvt;
    }

    public @NotNull UUID getId() {
        return id;
    }

    public @NotNull Location getLocation() {
        return location;
    }

    public @NotNull List<Line<?>> getLines() {
        return lines;
    }

    protected @NotNull BlockLine blockLine(@NotNull Function<@NotNull Player, @NotNull ItemStack> blockSupplier) {
        BlockLine line = new BlockLine(this, blockSupplier);
        lines.add(line);
        return line;
    }

    protected @NotNull ItemLine itemLine(@NotNull Function<@NotNull Player, @NotNull ItemStack> itemSupplier) {
        ItemLine line = new ItemLine(this, itemSupplier);
        lines.add(line);
        return line;
    }

    protected @NotNull TextLine textLine(@NotNull Function<@NotNull Player, @NotNull String> textSupplier) {
        TextLine line = new TextLine(this, textSupplier);
        lines.add(line);
        return line;
    }

    @ApiStatus.Experimental
    protected @NotNull DisplayTextLine displayTextLine(@NotNull Function<@NotNull Player, @NotNull String> textSupplier) {
        DisplayTextLine line = new DisplayTextLine(this, textSupplier);
        lines.add(line);
        return line;
    }

    @ApiStatus.Experimental
    protected @NotNull DisplayBlockLine displayBlockLine(@NotNull Function<@NotNull Player, @NotNull Material> materialSupplier) {
        DisplayBlockLine line = new DisplayBlockLine(this, materialSupplier);
        lines.add(line);
        return line;
    }

    @ApiStatus.Experimental
    protected @NotNull InteractionLine interactionLine() {
        InteractionLine line = new InteractionLine(this);
        lines.add(line);
        return line;
    }

    public void teleport(@NotNull Location to) {
        this.location = to.clone();

        // update the location of all lines
        pvt.updateLinesLocation();

        // send teleport packets to all players seeing this hologram
        for (Player seeingPlayer : pvt.getSeeingPlayers()) {
            for (Line<?> line : lines) {

                WrapperPlayServerEntityTeleport packet = new WrapperPlayServerEntityTeleport(
                        line.getEntityID(),
                        SpigotConversionUtil.fromBukkitLocation(line.getLocation()),
                        false
                );
                PacketEvents.getAPI().getPlayerManager().sendPacket(seeingPlayer, packet);
            }
        }
    }

    public boolean isShownFor(@NotNull Player player) {
        return pvt.getSeeingPlayers().contains(player);
    }

    public <T extends Hologram> void show(@NotNull IHologramPool<T> pool) {
        if (pool.getHolograms().stream().anyMatch(h -> h.getId().equals(this.id))) {
            throw new KeyAlreadyExistsException(this.id);
        }
        pool.getHolograms().add((T) this);
    }

    public void show(@NotNull Player player) {
        if (!loaded) {
            if(lines.isEmpty()) {
                throw new IllegalStateException("Cannot show hologram with no lines.");
            }
            pvt.updateLinesLocation();
            loaded = true;
        }

        pvt.getSeeingPlayers().add(player);
        for (Line<?> line : lines) {
            (line).show(player);
        }

        if (pvt.getShowEvent() != null) {
            pvt.getShowEvent().onShow(player);
        }
    }

    public void hide(@NotNull Player player) {
        for (Line<?> line : lines) {
            (line).hide(player);
        }
        pvt.getSeeingPlayers().remove(player);

        if (pvt.getHideEvent() != null) {
            pvt.getHideEvent().onHide(player);
        }
    }

    public void hide(@NotNull IHologramPool<?> pool) {
        boolean removed = pool.getHolograms().remove(this);
        if (removed) {
            for (Player player : pvt.getSeeingPlayers()) {
                hide(player);
            }
        }
    }

    public void updateLines() {
        for (Line<?> line : lines) {
            line.updateAll();
        }
    }

    public void replaceLines(@NotNull List<? extends Line<?>> newLines) {
        // hide current lines from all players
        for (Player player : pvt.getSeeingPlayers()) {
            for (Line<?> line : lines) {
                line.hide(player);
            }
        }

        lines.clear();
        loaded = false;
        lines.addAll(newLines);
        pvt.updateLinesLocation();

        // show new lines to all players
        for (Player player : pvt.getSeeingPlayers()) {
            for (Line<?> line : lines) {
                line.show(player);
            }
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Hologram)) return false;
        Hologram hologram = (Hologram) o;
        return id.equals(hologram.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
}

File: holoeasy-core/src/main/kotlin/org/holoeasy/hologram/Hologram.kt
package org.holoeasy.hologram

import org.bukkit.Location
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.holoeasy.HoloEasy
import org.holoeasy.builder.TextLineModifiers
import org.holoeasy.line.*
import org.holoeasy.pool.IHologramPool
import org.holoeasy.pool.KeyAlreadyExistsException
import org.holoeasy.reactive.MutableState
import org.jetbrains.annotations.ApiStatus.Internal
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList


open class Hologram @JvmOverloads constructor(
    location: Location,
    showEvent: ShowEvent? = null,
    hideEvent: HideEvent? = null
) {

    @Internal
    val pvt = PrivateConfig(this, showEvent, hideEvent)
    val id = UUID.randomUUID()!!
    var location: Location = location
        private set
    private val stateVars = mutableListOf<MutableState<*>>()
    val lines: MutableList<Line<*>> = CopyOnWriteArrayList() // writes are slow and Iterators are fast and consistent.

    // if is first loaded
    private var loaded = false

    open fun loader(): HologramLoader = HologramLoader.TEXT_BLOCK_STANDARD

    protected fun <T : Any> mutableStateOf(value: T): MutableState<T> {
        val state = MutableState(value)
        stateVars.add(state)
        return state
    }

    protected fun blockLine(item: ItemStack): Line<ItemStack> {
        val line = BlockLine(hologram = this, obj = item)
        lines.add(line)
        return line
    }

    protected fun itemLine(item: ItemStack): Line<ItemStack> {
        val line = ItemLine(hologram = this, obj = item)
        lines.add(line)
        return line
    }

    @JvmOverloads
    protected fun textLine(text: String, modifiers: TextLineModifiers = TextLineModifiers()): Line<String> {
        val line = if (modifiers.clickable) {
            if (modifiers.clickableWithoutPool) {
                val clickableTextLine =
                    ClickableTextLine(
                        hologram = this,
                        text = text,
                        args = modifiers.args,
                        clickable = false,
                        minHitDistance = modifiers.minHitDistance,
                        maxHitDistance = modifiers.maxHitDistance,
                    )
                modifiers.clickEvent?.let { clickableTextLine.clickEvent = it }
                clickableTextLine
            } else {
                val textLine = TextLine(hologram = this, text = text, clickable = true, args = modifiers.args)
                modifiers.clickEvent?.let { textLine.clickEvent = it }
                textLine
            }
        } else {
            TextLine(hologram = this, text = text, clickable = false, args = modifiers.args)
        }
        lines.add(line)
        return line
    }

    fun teleport(to: Location) {
        this.location = to.clone()
        loader().teleport(this)
    }

    fun isShownFor(player: Player): Boolean {
        return pvt.seeingPlayers.contains(player)
    }

    @JvmOverloads
    fun show(pool: IHologramPool<*> = HoloEasy.standardPool) {
        if (pool.holograms.any { it.id == this.id }) {
            throw KeyAlreadyExistsException(this.id)
        }
        (pool.holograms as MutableSet<Hologram>).add(this);
    }

    fun show(player: Player) {
        if (!loaded) {
            pvt.load()
            loaded = true
        }

        pvt.seeingPlayers.add(player)
        for (line in lines) {
            (line as LineImpl).show(player)
        }

        pvt.showEvent?.onShow(player)
    }

    fun hide(player: Player) {
        for (line in lines) {
            (line as LineImpl).hide(player)
        }
        pvt.seeingPlayers.remove(player)

        pvt.hideEvent?.onHide(player)
    }

    @JvmOverloads
    fun hide(pool: IHologramPool<*> = HoloEasy.standardPool) {
        // if removed
        val removed = (pool.holograms as MutableSet<Hologram>).remove(this)
        if (removed) {
            for (player in pvt.seeingPlayers) {
                hide(player)
            }
        }
    }


    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Hologram

        return id == other.id
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }

    fun serialize(): MutableMap<String, Any> {
        val result = LinkedHashMap<String, Any>()
        result["location"] = location
        result["lines"] = lines
            .map { mapOf("type" to it.type.name, "value" to it.value) }
        result["stateVars"] = stateVars.map { it.get() }
        return result
    }

    companion object {

        @JvmStatic
        fun <T : Hologram> deserialize(args: Map<String, Any>, clazz: Class<T>): T {
            val location = args["location"] as Location
            val lines = args["lines"] as List<Map<String, Any>>
            val stateVars = args["stateVars"] as List<Any>

            val hologram = clazz
                .getDeclaredConstructor(Location::class.java)
                .newInstance(location) as T

            for (i in lines.indices) {
                val type = LineImpl.Type.valueOf(lines[i]["type"] as String)
                val value = lines[i]["value"]

                when (type) {
                    LineImpl.Type.EXTERNAL -> {
                        val hologramLine = hologram.lines[i] as LineImpl<Any>
                        hologramLine.value = value as Any
                    }

                    LineImpl.Type.TEXT_LINE, LineImpl.Type.CLICKABLE_TEXT_LINE -> {
                        val hologramLine = hologram.lines[i] as TextLine
                        hologramLine.value = value as String
                    }

                    LineImpl.Type.ITEM_LINE, LineImpl.Type.BLOCK_LINE -> {
                        val hologramLine = hologram.lines[i] as LineImpl<ItemStack>
                        hologramLine.value = value as ItemStack
                    }
                }
            }
            for (i in stateVars.indices) {
                val hologramState = hologram.stateVars[i] as MutableState<Any>
                hologramState.set(stateVars[i])
            }


            return hologram
        }

        inline fun <reified T : Hologram> deserialize(args: Map<String, Any>): T {
            return deserialize(args, T::class.java)
        }

    }
}
